subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -524,19 +524,28 @@ public ServiceProvider getService(LookupContext context, TypeSpec serviceType) {
             if (providers == null) {
                 return null;
             }
-            List<ServiceProvider> candidates = new ArrayList<ServiceProvider>();
+            ServiceProvider singleCandidate = null;
+            List<ServiceProvider> candidates = null;
             for (Provider provider : providers) {
                 ServiceProvider service = provider.getService(context, serviceType);
                 if (service != null) {
-                    candidates.add(service);
+                    if (singleCandidate == null) {
+                        singleCandidate = service;
+                    } else {
+                        if (candidates == null) {
+                            candidates = new ArrayList<ServiceProvider>(2);
+                            candidates.add(singleCandidate);
+                        }
+                        candidates.add(service);
+                    }
                 }
             }
 
-            if (candidates.size() == 0) {
+            if (candidates == null && singleCandidate == null) {
```

Comment:
Okay, so you did see this array list in the profiler? I'm asking because hotspot is normally good at optimizing intermediate objects away through escape analysis.